File: missing-number/solution268.go
// https://leetcode.com/problems/missing-number/

package main

import (
	"fmt"
)

// T: O(n), S: O(1)
func missingNumberSum(nums []int) int {
	n := len(nums)
	expectedSum := n * (n + 1) / 2
	actualSum := 0
	for _, num := range nums {
		actualSum += num
	}
	return expectedSum - actualSum
}

// T: O(n), S: O(1)
func missingNumberXor(nums []int) int {
	missing := len(nums)
	for i, num := range nums {
		missing ^= i ^ num
	}
	return missing
}

// T: O(n), S: O(n)
func missingNumberBitSet(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	set := make([]bool, n+1)
	for _, num := range nums {
		set[num] = true
	}
	for i := 0; i <= n; i++ {
		if !set[i] {
			return i
		}
	}
	return -1
}

func main() {
	nums1 := []int{3, 0, 1}
	nums2 := []int{0, 1}
	nums3 := []int{9, 6, 4, 2, 3, 5, 7, 0, 1}

	fmt.Println(missingNumberSum(nums1))    // 2
	fmt.Println(missingNumberXor(nums1))    // 2
	fmt.Println(missingNumberBitSet(nums1)) // 2

	fmt.Println(missingNumberSum(nums2))    // 2
	fmt.Println(missingNumberXor(nums2))    // 2
	fmt.Println(missingNumberBitSet(nums2)) // 2

	fmt.Println(missingNumberSum(nums3))    // 8
	fmt.Println(missingNumberXor(nums3))    // 8
	fmt.Println(missingNumberBitSet(nums3)) // 8
}

File: missing-number/Solution268.java
// https://leetcode.com/problems/missing-number/

import java.util.*;

class Solution268 {

    // T: O(n), S: O(1)
    public int missingNumberSum(int[] nums) {
        int n = nums.length;
        int expectedSum = n * (n + 1) / 2;
        int actualSum = 0;
        for (int num : nums) {
            actualSum += num;
        }
        return expectedSum - actualSum;
    }

    // T: O(n), S: O(1)
    public int missingNumberXor(int[] nums) {
        int missing = nums.length;
        for (int i = 0; i < nums.length; i++) {
            missing ^= i ^ nums[i];
        }
        return missing;
    }

    // T: O(n), S: O(n)
    public int missingNumberBitSet(int[] nums) {
        int n = nums.length;
        if (n == 0) {
            return 0;
        }
        var set = new BitSet(n + 1);
        for (int num : nums) {
            set.set(num);
        }
        for (int i = 0; i < n + 1; i++) {
            if (set.get(i) == false) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        var sol = new Solution268();
        var nums1 = new int[] { 3, 0, 1 };
        var nums2 = new int[] { 0, 1 };
        var nums3 = new int[] { 9, 6, 4, 2, 3, 5, 7, 0, 1 };

        System.out.println(sol.missingNumberSum(nums1)); // 2
        System.out.println(sol.missingNumberXor(nums1)); // 2
        System.out.println(sol.missingNumberBitSet(nums1)); // 2

        System.out.println(sol.missingNumberSum(nums2)); // 2
        System.out.println(sol.missingNumberXor(nums2)); // 2
        System.out.println(sol.missingNumberBitSet(nums2)); // 2

        System.out.println(sol.missingNumberSum(nums3)); // 8
        System.out.println(sol.missingNumberXor(nums3)); // 8
        System.out.println(sol.missingNumberBitSet(nums3)); // 8
    }

}
